### PR TITLE
Add Push-To-Talk feature. This closes #72.

### LINKED ISSUE #72: FR: need push-to-talk mode
Would be useful to be able to configure a keyboard shortcut for push-to-talk like with mumble.

### PR DESCRIPTION
**Pull request by sandr3j**
I implemented the requested Push-To-Talk feature. It was necessary to patch a library mousetrap-record licensed under Apache 2.0 license. It was also necessary to use an other angular-hotkey repository with a patched bug.

### CODE REVIEW

File: src/lib/mousetrap-record.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
+/**
+ * This extension allows you to record a sequence using Mousetrap.
```
Comment (ancorgs) [inline src/lib/mousetrap-record.js]: Mmm, I wonder if there is not a better way to include this file.

File: src/lib/mousetrap-record.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            return;
+        }
+
+        // remember this character if we're currently recording a sequence
+        if (e.type == 'keydown') {
+            if (character.length === 1 && _recordedCharacterKey) {
+                _recordCurrentCombo();
+            }
+
+            for (i = 0; i < modifiers.length; ++i) {
```
Comment (ancorgs) [inline src/lib/mousetrap-record.js]: This throws an error. There is a missing `var i;` at the beginning of the function.

**Comment by ancorgs** [comment]
Even after fixing the mentioned lack of declaration of `i` in `_handleKey`, the behavior looks buggy to me. Playing a little bit with it, turns out to be relatively easy to reach a situation in which it does not record the last pressed key, but the previous one.

File: src/app/components/feed/jh-pushtotalk-button.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<button ng-class="{'btn-danger': vm.hotkeyActive }" class="btn" ng-click="vm.click()" title="{{vm.titleText}}">
+  <span class="glyphicon glyphicon-bullhorn" aria-hidden="true">
+  </span>
+  <span ng-show="vm.showHotkey">
+    <kbd style="background-color:#f15a5c">
+        {{vm.hotkey}}
```
Comment (ancorgs) [inline src/app/components/feed/jh-pushtotalk-button.html]: Just a nickpicking detail. When the mode is activated and the key is displayed, the header grows a little bit, at least in Firefox.

File: src/app/components/room/action.service.js [hunk 4]
```diff
@@ -131,5 +132,12 @@
         feed.setEnabledChannel(type, true);
       }
     }
+
+    function setMainFeedChannel(type, boolval) {
```
Comment (ancorgs) [inline src/app/components/room/action.service.js]: Pretty non-intuitive name for this function. I know that `setEnabledMainFeedChannel` is ridiculous long, but I'm sure we can find a better alternative.

File: src/app/components/room/room.service.js [hunk 5]
```diff
@@ -416,6 +418,16 @@
       ActionService.toggleChannel(type, feed);
     }
 
+    function pushToTalk(keyevent){
```
Comment (ancorgs) [inline src/app/components/room/room.service.js]: Please, try to document new functions added to `RoomService`.

**Comment by ancorgs** [comment]
I like how the feature is implemented from the user interaction point of view. We just need to fix a couple of small bugs.

File: src/lib/mousetrap-record.js (hunk 2 above)
Comment (sandr3j) [inline src/lib/mousetrap-record.js]: You're right. On Chrome it worked without var i, and for Firefox I forgot to make the change persistent.

File: src/app/components/feed/jh-pushtotalk-button.html (hunk 3 above)
Comment (sandr3j) [inline src/app/components/feed/jh-pushtotalk-button.html]: Ah I didn't notice that. It grows by one or two pixels.

Comment (sandr3j) [inline src/app/components/feed/jh-pushtotalk-button.html]: You're right. I did now some stress testing and I see that it records the previous key. I also uncovered a second bug in my function abortKey, I need to add `if(_recordedSequenceCallback !== null)` because if aborting very fast before it finished triggering all callbacks it throws an error.

File: src/lib/mousetrap-record.js (hunk 1 above)
Comment (sandr3j) [inline src/lib/mousetrap-record.js]: We could move it to app/ and exclude in gulp or fix all annoying jslint errors. If the library is complete garbage we could use this: https://dmauro.github.io/Keypress/
But it's a much bigger library in comparison to this small Mousetrap plugin.

File: src/app/components/room/action.service.js (hunk 4 above)
Comment (sandr3j) [inline src/app/components/room/action.service.js]: `setMedia` or `enableMedia` sounds for me more intuitive.

File: src/app/components/room/room.service.js (hunk 5 above)
Comment (sandr3j) [inline src/app/components/room/room.service.js]: Yes, a small description like for `sendStatus` sounds useful.

**Comment by sandr3j** [comment]
Yop, the most annoying bug I see is that the previous key can get recorded. I will try to fix that.

**Comment by sandr3j** [comment]
@ancorgs I pushed now bug fixes. Please review :)

File: src/lib/mousetrap-record.js (hunk 1 above)
Comment (ancorgs) [inline src/lib/mousetrap-record.js]: My concern was not about the location of the file. I was thinking about including Mousetrap through bower.json and either contributing the fixes to the upstream Mousetrap either maintaining our own fork with the fixes/enhancement. 

Take into account that currently Mousetrap lives here https://github.com/mousetrap-js/mousetrap (not in the original ccampbell repo). They look very welcoming to fixes and improvements.

So in short, don't copy&paste&modify code but collaborate with the maintainers.

Comment (sandr3j) [inline src/lib/mousetrap-record.js]: The thing is ccampbell repo seems not to be maintained and I simply missed this fresh fork. The best thing they updated [the plugin](https://github.com/mousetrap-js/mousetrap/blob/master/plugins/record/mousetrap-record.js) with the features we need!

They implemented `stopRecord` function and I can pass `recordSequence=false` to remove the delay after the first key is recorded. I will reimplement key recoding using their library and will add their repository to bower.

**Comment by sandr3j** [comment]
@ancorgs Hey, I thought we should complete this feature regardless of GSoC. The mousetrap record plugin in the mousetrap-js repository was still buggy. I fixed it and [commited my fix to mousetrap-js repository](https://github.com/mousetrap-js/mousetrap/commit/8236fbda90e2c50dfbfbe4e3f26258ad3a873968). 
The plugin should work now snappy and bug free. Please review again if everything is ok so this can get merged :)

**Comment by ancorgs** [comment]
The branch has conflicts. Looks like you need to rebase

**Comment by sandr3j** [comment]
I have taken care of the conflict, beautified the code and added preventDefault for hotkey.

**Comment by ancorgs** [comment]
I have very limited internet access these days. I'll check (and surely merge) tomorrow or during weekend.

**Comment by ancorgs** [comment]
Mmm, I'm having some problems testing this. Looks like gulp is not including the mousetrap plugins into vendor.js. Still researching

File: src/index.html [hunk 6]
```diff
@@ -43,6 +43,7 @@
      <link rel="stylesheet" href="bower_components/angular-gridster/dist/angular-gridster.min.css"/>
      <script src="bower_components/javascript-detect-element-resize/jquery.resize.js"></script>
      <script src="bower_components/angular-gridster/dist/angular-gridster.min.js"></script>
+     <script src="bower_components/mousetrap/plugins/record/mousetrap-record.js"></script>
```
Comment (ancorgs) [inline src/index.html]: This approach only works with `gulp serve` but not with `gulp build`, so it's not valid.

The three preceding lines (including angular-gridster and jquery.resize) are actually garbage and should also be removed because that code is already included in vendor.js and vendor.css.

We need a clean way to make sure mousetrap-record is included in vendor.js as well.

**Comment by sandr3j** [comment]
Hey, cool that you finally could review my commit! Sorry, I didn't want to cause any headache. I only tested using `gulp serve`. It was very tempting to include it that way and there are even [blogs](http://zellwk.com/blog/bower/#using-a-library-installed-with-bower) which recommend this. Anyway it seems like [wiredep bower overrides](https://github.com/taptapship/wiredep#bower-overrides) is the way to go here. I tested now `gulp build` and there are no problems anymore. I also removed the unused garbage. Please check if this issue is fixed.

**Comment by ancorgs** [comment]
It works nicely now at a first sight. But it seems to have some small issues according to the comments on https://github.com/jangouts/jangouts/issues/72. I also agree with those comments that the setting should be remembered, pretty much like we already do with the layout, the room, the nickname and so on.

This feature is taking quite some time to shape up, but I really really like the result so far. :smile:

**Comment by sandr3j** [comment]
Somehow I spent on this more time then I want to admit. So long that I wrote javascript which freezes Firefox and crashes Chrome. I created an infinite recursive call and Chrome began leaking massive memory until the tab crashed (which I confirmed with gdb). It looks like it only happens while a video feed is streamed. The same code freezes Firefox. But isolating the issue and document it on the bug tracker is not that fun :S

Anyway there is the workaround for the superkey for Firefox and Chrome. And then I uncovered another bug for key combinations. When you release pressed keys in the wrong order the keyup event is fired so I wrote for this a workaround too. Somehow the workarounds are too big and hackish so we might consider to remove support for key combinations in favor of better maintainability. Somehow I didn't expect that these hotkey libraries are so buggy...

**Comment by ancorgs** [comment]
I know that feeling. :(

Just remove support and move forward.

I don't think a "restricted support" message is needed either.  Simply document somewhere in the code the reasons that leaded to the decision. At some point (not as part of this PR) we should create a "help" option (maybe improving the current "keyword shortcuts" one) and there we can explain something to the user. But as said, not now.

**Comment by sandr3j** [comment]
I removed now support for key combinations. Let me know if there is something else that need to be fixed.

File: src/app/components/feed/jh-pushtotalk.directive.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    return {
+      restrict: 'E',
+      templateUrl: 'app/components/feed/jh-pushtotalk-button.html',
+      controllerAs: 'vm',
+      controller: jhPushToTalkButtonCtrl,
+      scope: {}
+    };
+
+    function jhPushToTalkButtonCtrl() {
+      var titleText1 = "Set Push-to-talk hotkey or combination";
```
Comment (ancorgs) [inline src/app/components/feed/jh-pushtotalk.directive.js]: It still says "or combination" here.

File: src/app/components/feed/jh-pushtotalk.directive.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            combo: key,
+            description: 'Push-to-talk',
+            callback: pttCallback,
+            action: 'keyup'
+          });
+
+          vm.hotkey = key;
+          vm.showHotkey = true;
+          vm.titleText = titleText2;
+          UserService.setSetting("lastHotkey", key);
```
Comment (ancorgs) [inline src/app/components/feed/jh-pushtotalk.directive.js]: This persists the key when push-to-talk is set. But it does not "forget" the setting when push-to-talk is disabled. That's unexpected. If I enable push-to-talk once, I will have it automatically enabled on every subsequent login, even if I disable it every time.

File: src/app/components/feed/jh-pushtotalk-button.html [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<button ng-class="{'btn-danger': vm.hotkeyActive }" class="btn" ng-click="vm.click($event)" title="{{vm.titleText}}">
+  <span class="glyphicon glyphicon-bullhorn" aria-hidden="true">
+  </span>
+  <span ng-show="vm.showHotkey">
+    <kbd style="background-color:#f15a5c;line-height:16px;">
```
Comment (ancorgs) [inline src/app/components/feed/jh-pushtotalk-button.html]: This still produces de 1px jump in Firefox. Better use `line-height:1em`, which works in Chrom(e/ium) and Firefox.

Comment (sandr3j) [inline src/app/components/feed/jh-pushtotalk-button.html]: Actually I thought the "in the other news"  phrase was just a joke because I don't have this problem  using Firefox 45.

**Comment by sandr3j** [comment]
I fixed now what you mentioned.

**Comment by ancorgs** [comment]
Great job! Thanks a lot!
